libraries/AP_Baro/AP_Baro_ICM20789.cpp

```diff
@@ -0,0 +1,318 @@
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+    return sensor;
+}
+
+
+bool AP_Baro_ICM20789::spi_init(void)
+{
+    dev_icm->set_read_flag(0x80);
+
+    dev_icm->set_speed(AP_HAL::Device::SPEED_LOW);
+    uint8_t whoami = 0;
+    uint8_t v;
+        
+    dev_icm->read_registers(0x6A, &v, 1);
+    dev_icm->write_register(0x6B, 0x01);
+    
+    hal.scheduler->delay(1);
+    dev_icm->write_register(0x6A, 0x10);
+    dev_icm->write_register(0x6B, 0x41);
+    
+    hal.scheduler->delay(1);
+    dev_icm->write_register(0x6B, 0x01);
+    
+    hal.scheduler->delay(1);
+    dev_icm->write_register(0x23, 0x00);
+    dev_icm->write_register(0x6B, 0x41);
+    
+    dev_icm->read_registers(0x75, &whoami, 1);
+    
+    // wait for sensor to settle
+    hal.scheduler->delay(100);
+    
+    dev_icm->read_registers(0x75, &whoami, 1);
+    
+    dev_icm->write_register(0x37, 0x00);
+    dev_icm->write_register(0x6A, 0x10);
+    return true;
+}
+
+bool AP_Baro_ICM20789::init()
+{
+    if (!dev) {
+        return false;
+    }
+
+    debug("Looking for 20789 baro\n");
+    
+    if (!dev->get_semaphore()->take(0)) {
+        AP_HAL::panic("PANIC: AP_Baro_ICM20789: failed to take serial semaphore for init");
+    }
+
+#ifdef HAL_INS_MPU60x0_NAME
+    /*
+      Pressure sensor data can be accessed in the following mode:
+      Bypass Mode: Set register INT_PIN_CFG (Address: 55 (Decimal); 37 (Hex)) bit 1 to value 1 
+      and I2C_MST_EN bit is 0 Address: 106 (Decimal); 6A (Hex)). 
+      Pressure sensor data can then be accessed using the procedure described in Section 10.
+    */
+    debug("Setting up IMU\n");
+    dev_icm = hal.spi->get_device(HAL_INS_MPU60x0_NAME);
+
+    if (!dev_icm->get_semaphore()->take(0)) {
+        AP_HAL::panic("PANIC: AP_Baro_ICM20789: failed to take serial semaphore ICM");        
+    }
+
+    if (!spi_init()) {
+        debug("ICM20789: failed to initialise SPI device\n");
+        goto failed;
+    }
+
+    dev_icm->get_semaphore()->give();
+#endif // HAL_INS_MPU60x0_NAME
+    
+    if (!send_cmd16(CMD_SOFT_RESET)) {
+        debug("ICM20789: reset failed\n");
+        goto failed;
+    }
+
+    // wait for sensor to settle
+    hal.scheduler->delay(10);
+    
+    if (!read_calibration_data()) {
+        debug("ICM20789: read_calibration_data failed\n");
+        goto failed;
+    }
+        
+    // start a reading
+    if (!send_cmd16(CMD_READ_PT)) {
+        debug("ICM20789: start read failed\n");
+        goto failed;
+    }
+
+    dev->set_retries(0);
+
+    instance = _frontend.register_sensor();
+
+    dev->get_semaphore()->give();
+
+    debug("ICM20789: startup OK\n");
+    
+    // use 10ms to ensure we don't lose samples, with max lag of 10ms
+    dev->register_periodic_callback(CONVERSION_INTERVAL/2, FUNCTOR_BIND_MEMBER(&AP_Baro_ICM20789::timer, void));
+        
+    return true;
+
+ failed:
+    dev_icm->get_semaphore()->give();
```

Comment:
What if HAL_INS_MPU60x0_NAME isn't defined? :slightly_smiling_face:

Reply:
Going to change things so we don't take the semaphore if the name isn't defined.